pkg/dns/treecache.go

```diff
@@ -130,32 +90,38 @@ func (cache *TreeCache) GetSubCache(path ...string) *TreeCache {
 	return childCache
 }
 
-func (cache *TreeCache) SetSubCache(key string, subCache *TreeCache, path ...string) {
-	cache.m.Lock()
-	defer cache.m.Unlock()
+func (cache *TreeCache) setSubCache(key string, subCache *TreeCache, path ...string) {
+	if cache.rwLock != nil {
+		cache.rwLock.Lock()
+		defer cache.rwLock.Unlock()
+	}
 	node := cache.ensureChildNode(path...)
 	node.ChildNodes[key] = subCache
 }
 
-func (cache *TreeCache) GetEntry(key string, path ...string) (interface{}, bool) {
-	cache.m.RLock()
-	defer cache.m.RUnlock()
-	childNode := cache.GetSubCache(path...)
+func (cache *TreeCache) getEntry(key string, path ...string) (interface{}, bool) {
+	if cache.rwLock != nil {
+		cache.rwLock.RLock()
+		defer cache.rwLock.RUnlock()
+	}
+	childNode := cache.getSubCache(path...)
 	val, ok := childNode.Entries[key]
 	return val, ok
 }
 
-func (cache *TreeCache) GetValuesForPathWithRegex(path ...string) []interface{} {
-	cache.m.RLock()
-	defer cache.m.RUnlock()
+func (cache *TreeCache) getValuesForPathWithWildcards(path ...string) []interface{} {
```

Comment:
I'd rather do the locking outside of this function and document that this assumes it is called with the lock held.